rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp

```diff
@@ -75,7 +76,9 @@ class ParameterMutationRecursionGuard
 };
 
 /// Implementation of the NodeParameters part of the Node API.
-class NodeParameters : public NodeParametersInterface
+class NodeParameters
+  : public NodeParametersInterface,
+  public std::enable_shared_from_this<NodeParameters>
```

Comment:
This is not an ideal change, is it possible you don't actually need shared ownership of this? Would it be better to return a shared_ptr of the callback, and keep a weak ptr in this class?

Reply:
I tried to avoid the possibility of UB:
https://github.com/ros2/rclcpp/blob/c7d01d7bf3a6f886577689877205ffb08ff27b42/rclcpp/src/rclcpp/clock.cpp#L128-L140

If not, I can't remove the pointer from the container in the deleter.
I have two possible work-arounds:
- Do what `clock` does, and suppose people will write the program in a way that the callback goes out of scope before the interface (so, catching `this` and not `shared_from_this` is ok). This usually happens, when inheriting from `rclcpp::Node`.
- Let the callback go out of scope, and do not delete it from the container in its deleter (use the default deleter). Store weak ptrs (and not raw ptrs), in the container. The callbacks are removed from the container if they are found to be `nullptr` when trying to call them (or, they aren't removed at all, just skipped).

Reply:
I think we are always creating a shared pointer of `NodeParameters`.
So inheriting from `std::enable_shared_from_this` shouldn't be a problem (I should run full ci for confirming it).

Reply:
I think both of the alternatives you suggested are better than having enable shared from this (personally). The first one is more dangerous for inexperienced users that make a mistake, but the second one is pretty safe, it just incurs the overhead of a shared pointer (versus a unique pointer) for each callback handle returned. Personally I'd go with the latter options. That mimics how callback groups and pub/sub/service's are handled within the node.

> I think we are always creating a shared pointer of NodeParameters.
So inheriting from std::enable_shared_from_this shouldn't be a problem (I should run full ci for confirming it).

That's true, but I feel that we're limiting our options (I'd like to eventually get the node interfaces as a non-shared pointer part of the node. Even if that might not be possible, this change takes us further from that goal, and I think it's not required.

Also it means that when the `NodeParameters` is destructed, all outstanding cleanup is done, and you don't have to worry about incomplete cleanup due to lingering callback handle objects.